modules/gapi/test/gapi_async_test.cpp

```diff
@@ -0,0 +1,262 @@
+struct Future: crtp_cast<crtp_final_t> {
+    std::future<void> f;
+
+    template<typename... Args >
+    void start_async(Args&&... args){
+        f = this->crtp_cast_(this)->async(std::forward<Args>(args)...);
+    }
+
+    void wait_for_result()
+    {
+        f.get();
+    }
+};
+
+
+template<typename crtp_final_t>
+struct AsyncCompiled  : crtp_cast<crtp_final_t>{
+
+    template<typename... Args>
+    auto async(Args&&... args) -> decltype(cv::gapi::wip::async(std::declval<cv::GCompiled&>(), std::forward<Args>(args)...)){
+        auto gcmpld = this->crtp_cast_(this)->compile();
+        return cv::gapi::wip::async(gcmpld, std::forward<Args>(args)...);
+    }
+};
+
+template<typename crtp_final_t>
+struct AsyncApply : crtp_cast<crtp_final_t> {
+
+    template<typename... Args>
+    auto async(Args&&... args) ->decltype(cv::gapi::wip::async_apply(std::declval<cv::GComputation&>(), std::forward<Args>(args)...)) {
+        return cv::gapi::wip::async_apply(this->crtp_cast_(this)->computation(), std::forward<Args>(args)..., this->crtp_cast_(this)->compile_args());
+    }
+};
+
+
+template<typename case_t>
+struct normal: case_t{};
+
+TYPED_TEST_CASE_P(normal);
+
+TYPED_TEST_P(normal, basic){
+    this->start_async(this->in_args(), this->out_args());
+    this->wait_for_result();
+
+    this->verify();
+}
+
+REGISTER_TYPED_TEST_CASE_P(normal,
+        basic
+);
+
+template<typename case_t>
+struct exception: case_t{};
+TYPED_TEST_CASE_P(exception);
+
+TYPED_TEST_P(exception, basic){
+    this->start_async(this->in_args(), this->out_args());
+    EXPECT_THROW(this->wait_for_result(), gthrow_exception);
+}
+
+REGISTER_TYPED_TEST_CASE_P(exception,
+        basic
+);
+
+template<typename compute_fixture_t,template <typename> class callback_or_future_t, template <typename> class compiled_or_apply_t>
+struct Case
+        : compute_fixture_t,
+          callback_or_future_t<Case<compute_fixture_t,callback_or_future_t,compiled_or_apply_t>>,
+          compiled_or_apply_t <Case<compute_fixture_t,callback_or_future_t,compiled_or_apply_t>>
+{};
+
+template<typename computation_t>
+using cases = ::testing::Types<
+            Case<computation_t, CallBack, AsyncCompiled>,
+            Case<computation_t, CallBack, AsyncApply>,
+            Case<computation_t, Future,   AsyncCompiled>,
+            Case<computation_t, Future,   AsyncApply>
+            >;
+INSTANTIATE_TYPED_TEST_CASE_P(AsyncAPINormalFlow_,        normal,     cases<SumOfSum2x2>);
+INSTANTIATE_TYPED_TEST_CASE_P(AsyncAPIExceptionHandling_, exception,  cases<ExceptionOnExecution>);
+
+
+TEST(AsyncAPI, DISABLED_Sample){
+    cv::GComputation self_mul([]{
+        cv::GMat in;
+        cv::GMat out = cv::gapi::mul(in, in);
+        return GComputation{in, out};
+    });
+
+    const cv::Size sz{2, 2};
+    cv::Mat in_mat{sz, CV_8U, cv::Scalar(1)};
+    cv::Mat out;
+
+    auto f = cv::gapi::wip::async_apply(self_mul,cv::gin(in_mat), cv::gout(out));
+    f.wait();
+}
```

Comment:
Why it is DISABLED BTW?

Reply:
enabled back

Reply:
No comments at all. What is the general idea behind these tests?